File: docs/bfs/1466_medium/reordered_routes.py
from collections import deque


def find_min_reorders(n: int, connections: list[list[int]]) -> int:
    src_nodes: list[list[int]] = [[] for _ in range(n)]  # Each node's source nodes.
    tgt_nodes: list[list[int]] = [[] for _ in range(n)]  # Each node's target nodes.

    visited: list[bool] = [False] * n

    for src_node, tgt_node in connections:
        src_nodes[tgt_node].append(src_node)
        tgt_nodes[src_node].append(tgt_node)

    min_reorders = 0
    queue: deque[int] = deque([0])  # Stores nodes.

    while queue:
        node = queue.popleft()
        visited[node] = True

        for src_node in src_nodes[node]:
            if not visited[src_node]:
                queue.append(src_node)

        for tgt_node in tgt_nodes[node]:
            if not visited[tgt_node]:
                min_reorders += 1
                queue.append(tgt_node)

    return min_reorders

File: docs/bfs/1466_medium/reordered_routes.cpp
#include <queue>
#include <vector>
using namespace std;

int find_min_reorders(int n, vector<vector<int>>& connections) {
    vector<vector<int>> srcNodes(n, vector<int>()), tgtNodes(n, vector<int>());
    vector<bool> visited(n, false);

    for (const auto& edge : connections) {
        int srcNode = edge[0], tgtNode = edge[1];

        srcNodes[tgtNode].push_back(srcNode);
        tgtNodes[srcNode].push_back(tgtNode);
    }

    int minReorders = 0;

    queue<int> queue; // Stores nodes.
    queue.push(0);

    while (!queue.empty()) {
        int node = queue.front();
        queue.pop();
        visited[node] = true;

        for (const auto& srcNode : srcNodes[node]) {
            if (!visited[srcNode])
                queue.push(srcNode);
        }

        for (const auto& tgtNode : tgtNodes[node]) {
            if (!visited[tgtNode]) {
                minReorders++;
                queue.push(tgtNode);
            }
        }
    }

    return minReorders;
}
